src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,830 @@
+  apply ball_congr,
+  intros x tx,
+  apply exists_unique_congr,
+  intro t,
+  rw ← iso.to_equiv_symm_fun,
+  rw equiv.eq_symm_apply,
+  split,
+  { intros q,
+    ext Y f hf,
+    simpa [first_obj_eq_family, fork_map] using q _ _ },
+  { intros q Y f hf,
+    rw ← q,
+    simp [first_obj_eq_family, fork_map] }
+end
+
+end sieve
+
+/-!
+This section establishes the equivalence between the sheaf condition of
+https://stacks.math.columbia.edu/tag/00VM and the definition of `is_sheaf_for`.
+-/
+namespace presieve
+
+variables [has_pullbacks C]
+
+/--
+The rightmost object of the fork diagram of https://stacks.math.columbia.edu/tag/00VM, which
+contains the data used to check a family of elements for a presieve is compatible.
+-/
+def second_obj : Type v :=
+∏ (λ (fg : (Σ Y, {f : Y ⟶ X // R f}) × (Σ Z, {g : Z ⟶ X // R g})),
+  P.obj (op (pullback fg.1.2.1 fg.2.2.1)))
+
+/-- The map `pr₀*` of https://stacks.math.columbia.edu/tag/00VL. -/
+def first_map : first_obj P R ⟶ second_obj P R :=
+pi.lift (λ fg, pi.π _ _ ≫ P.map pullback.fst.op)
+
+instance : inhabited (second_obj P (⊥ : presieve X)) := ⟨first_map _ _ (default _)⟩
+
+/-- The map `pr₁*` of https://stacks.math.columbia.edu/tag/00VL. -/
+def second_map : first_obj P R ⟶ second_obj P R :=
+pi.lift (λ fg, pi.π _ _ ≫ P.map pullback.snd.op)
+
+lemma w : fork_map P R ≫ first_map P R = fork_map P R ≫ second_map P R :=
+begin
+  apply limit.hom_ext,
+  rintro ⟨⟨Y, f, hf⟩, ⟨Z, g, hg⟩⟩,
+  simp only [first_map, second_map, fork_map],
+  simp only [limit.lift_π, limit.lift_π_assoc, assoc, fan.mk_π_app, subtype.coe_mk,
+             subtype.val_eq_coe],
+  rw [← P.map_comp, ← op_comp, pullback.condition],
+  simp,
+end
+
+/--
+The family of elements given by `x : first_obj P S` is compatible iff `first_map` and `second_map`
+map it to the same point.
+-/
+lemma compatible_iff (x : first_obj P R) :
+  ((first_obj_eq_family P R).hom x).compatible ↔ first_map P R x = second_map P R x :=
+begin
+  rw presieve.pullback_compatible_iff,
+  split,
+  { intro t,
+    ext ⟨⟨Y, f, hf⟩, Z, g, hg⟩,
+    simpa [first_map, second_map] using t hf hg },
+  { intros t Y Z f g hf hg,
+    rw types.limit_ext_iff at t,
+    simpa [first_map, second_map] using t ⟨⟨Y, f, hf⟩, Z, g, hg⟩ }
+end
+
+/--
+`P` is a sheaf for `R`, iff the fork given by `w` is an equalizer.
+See https://stacks.math.columbia.edu/tag/00VM.
+-/
+lemma sheaf_condition :
+  R.is_sheaf_for P ↔ nonempty (is_limit (fork.of_ι _ (w P R))) :=
+begin
+  rw types.type_equalizer_iff_unique,
+  erw ← equiv.forall_congr_left (first_obj_eq_family P R).to_equiv.symm,
+  simp_rw [← compatible_iff, ← iso.to_equiv_fun, equiv.apply_symm_apply],
+  apply ball_congr,
+  intros x hx,
+  apply exists_unique_congr,
+  intros t,
+  rw equiv.eq_symm_apply,
+  split,
+  { intros q,
+    ext Y f hf,
+    simpa [fork_map] using q _ _ },
+  { intros q Y f hf,
+    rw ← q,
+    simp [fork_map] }
+end
+
+end presieve
+end equalizer
```

Comment:
Is there a convention for spaces between these `end`s? Not sure

Reply:
At the end of the file, I think it's common to just close everything without white lines between them. Doesn't matter too much, I would say.